### PR TITLE
[WIP] Extended CharacterForm, added CRUM (a plugin that may make some exist…

### PR DESCRIPTION
**Pull request by trhr**
…ing code redundant/easier to maintain)

#### Brief overview of PR changes/additions

- This ticks off one documentation-only issue from #533 .
- Overall, the goal of this is permission-based front-end character creation and management.

#### Motivation for adding to Evennia
I'm digging into #533; I wanted to get this little piece done before jumping into Admin ModelForms. 

#### Other info (issues closed, discussion etc)
Check out the crum library, as this PR adds it to requirements.txt. A lot of the frontend work from here till the end of time will need to access self.request.user in a function that doesn't pass "requests," and this keeps all that code clean and nice (as you know, trying to like pop a kwarg through multiple levels of super() inheritance is pretty messy.)

### CODE REVIEW

**Comment by trhr** [comment]
A bunch of #533 stuff is progressing nicely. I'm sure this will throw all kinds of bugs/errors if you try to mess with anything that's NOT a character object; I haven't finished all those bits yet.

- On the object's Edit view in the django admin, a field db_account is present. Changing the account here supports multisession modes (0, 1). Last-puppet gets set, existing sessions puppet their new characters. Multisession (2) works as well *except* when locks would prevent it. Maybe this is "working as intended" and requires some elbow grease by an admin to fix their locks, rather than inadvertently allowing any character to puppet any other character just because it gets set so in the admin.

- Added a new util, which is still in progress; a get_all_cmdsets function

- Started futzing around with some ideas for better typeclass/cmdset display. The Evennia core can't read the mygame dir, so I think that's why nothing custom is being added to either the typeclasses select or the commandset select. Any ideas?

**Comment by trhr** [comment]
- The superuser flag sits in the "web only" section. This is the only setting that does matter in-game in the same way as on the web, so it should probably be described somehow in that header (this is just a documentation issue).
- For MULTISESSION_MODE=0,1, one should be able to tie a Character to the Player by setting the Object's db_player foreignfield. In MULTISESSION_MODE=2, one might instead want to assign the character to the _playable_characters Attribute on the Player. A custom form is probably needed for this linking.
- There should probably be a way to connect a Player to a Character in the admin, for MULTISESSION_MODE=0,1 to not need to use @ic to connect.

Those are done.

- Possibly there should be some sort of list of suggested Object cmdsets so one doesn't need to remember and write the path every time. This could be a help text or if possible some selector based on settings.CMDSET_*. Maybe one could consider some sort of clickable buttons that fill in some standard presets ...?

This seems sketchy, given the work on get_all_cmdsets(). We could very easily present defaults, but that's not particularly useful.

- When setting a cmdset on a Character in the admin, the cmdset seems to be set to be visible outside the character. The result is that you get multi-command errors when executing commands in the same room as the new Character.

Surely this has been fixed by now? Steps to duplicate?

- There should probably be some help info or preset for how to set the puppet: lock to allow a Player to puppet a given Character, such as puppet:id(<player id>).

Belongs in docs, surely.

**Comment by Griatch** [comment]
@trhr Just to be clear - this is WIP PR, right? Please mark it as WIP in the title so it's clear it's still being worked on. Sounds great overall! The `get_all_cmdsets` function sounds interesting; I suppose it would be possible to fetch default cmdsets on objects (because they'd be stored on the object), but it'd be pretty impractical (or even impossible) to find all of them. 

>> Possibly there should be some sort of list of suggested Object cmdsets so one doesn't need to remember and write the path every time. This could be a help text or if possible some selector based on settings.CMDSET_*. Maybe one could consider some sort of clickable buttons that fill in some standard presets ...?

> This seems sketchy, given the work on get_all_cmdsets(). We could very easily present defaults, but that's not particularly useful.

Implementation may be tricky, but I don't see why the proposed functionality is sketchy. Manually having to enter paths is always going to be error-prone, and if the cmdset path is set wrong you can make the object unusable very easily. That said, it's clear that there is no central repository of cmdset paths, so presenting a list of the most common ones from settings (charactercmdset, accountcmdset etc) may be a good start at least? 

>> When setting a cmdset on a Character in the admin, the cmdset seems to be set to be visible outside the character. The result is that you get multi-command errors when executing commands in the same room as the new Character.

> Surely this has been fixed by now? Steps to duplicate?

This was not tested in a long while, but reproducing it (or not) should be simple - using the admin, create a character with the CharacterCmdSet on it, place it in a location. Then in-game, go to the location and enter `look`. If you get a multi-match error for the look command, it means the lock is not set on the new Character - because you are getting their `look` command as well as your own. 

>> There should probably be some help info or preset for how to set the puppet: lock to allow a Player to puppet a given Character, such as puppet:id().

> Belongs in docs, surely.

There is already ample docs for this; but this is the kind thing that is easy to forget if working in the admin, so some sort of text field about it is not a bad idea I don't think. 

> Started futzing around with some ideas for better typeclass/cmdset display. The Evennia core can't read the mygame dir, so I think that's why nothing custom is being added to either the typeclasses select or the commandset select. Any ideas?

I find it strange that importing e.g. settings.BASE_ACCOUNT_TYPECLASS in a running server would not access the game dir...

**Comment by Griatch** [comment]
Never mind, I'm a dummy and had evennia installed without the `-e` flag after rebuilding this computer :D Let me try again ...

**Comment by Griatch** [comment]
@trhr Bump! This is a very cool feature and it doesn't appear to be far from being wrapped up. Any plans on continuing work on it?

**Comment by trhr** [comment]
> @trhr Bump! This is a very cool feature and it doesn't appear to be far from being wrapped up. Any plans on continuing work on it?

no.

**Comment by Griatch** [comment]
Closing this, since no more work will be done on it, and the admin was reworked independently anyway. I did borrow the concept of the `get_all_cmdsets` helper function from this PR though, so thanks for that!
